README.md

```diff
@@ -0,0 +1,305 @@
+# sanctuary-argv
+
+The JavaScript ecosystem has many libraries for parsing command-line
+arguments; sanctuary-argv differs from these in a few important ways.
+
+## Philosophy
+
+sanctuary-argv is completely pure. It does not:
+
+  - reference global variables (such as `process.argv`);
+
+  - read from standard input;
+
+  - write to standard output or standard error;
+
+  - mutate its arguments; or
+
+  - throw exceptions.
+
+## Terminology
+
+    cut -s -d : -f 1 -- file1 file2 file3 = command
+        -s -d : -f 1 -- file1 file2 file3 = arguments
+        -s                                = flags = flag names
+           -d : -f 1                      = options
+           -d   -f                        = option names
+              :    1                      = option values
+                     --                   = separator
+                        file1 file2 file3 = positional arguments
+
+### Design decisions
+
+  - There are no implicit flags or options. If the command-line interface
+    is to support `--help`, for example, sanctuary-argv does not generate
+    the help text.
+
+  - Both long option names (e.g. `--extension`) and short option names
+    (e.g. `-x`) are supported.
+
+  - Short option names and long option names must be specified separately
+    (whereas [Commander.js][] supports `'-x, --extension <extension>'`).
+
+  - If a long option name has an associated value, the value must be
+    provided as a separate argument (e.g. `['--extension', '.js']`).
+    `--name=value`-style options are not supported.
+
+  - If a short option name has an associated value, the value must be
+    provided as a separate argument (e.g. `['-x', '.js']`).
+
+  - Multiple short flag names may be provided together, with up
+    to one short option name. `tar -zxvf filename` is short for
+    `tar -z -x -v -f filename`.
+
+[Commander.js]: https://github.com/tj/commander.js
+
+## Algorithm
+
+There are three inputs:
+
+  - a _specification_, a mapping from flag/option name to handler;
+  - an arbitrary _configuration value_; and
+  - an array of _arguments_.
+
+There are two outputs:
+
+  - a _configuration value_; and
+  - an array of _positional arguments_.
+
+The separator consists of two hyphens (`/^--$/`).
+
+Short flag/option names consist of a hyphen followed by a non-hyphen
+character (`/^-[^-]$/`).
+
+Long flag/option names consist of two hyphens followed by one or more
+characters (`/^--[^]+$/`).
+
+Arguments are processed from left to right. For each argument, there are
+three possibilities:
+
+  - The argument is the separator. Processing ends and the current
+    configuration value is returned along with the remaining arguments.
+
+  - The argument is a (short/long) flag/option name. It is looked up in the
+    specification. There are three possibilities:
```

Comment:
This is no longer accurate now that we permit `-xyz`. I will update this section.